portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -69,20 +77,42 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     String lastName =request.getParameter("lastName");
     String country = request.getParameter("country");
     String subject = request.getParameter("subject");
+    long timeStamp = System.currentTimeMillis();
 
     Entity message = new Entity("Response");
     message.setProperty("firstName", firstName);
     message.setProperty("lastName", lastName);
     message.setProperty("country", country);
     message.setProperty("subject",subject);
+    message.setProperty("timeStamp", timeStamp);
 
-    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     datastore.put(message);
-    
-
 
     response.sendRedirect("/contact.html");
 }
+
+/** Returns the choice entered by the user or -1 if the choice was invalid. */
+  private int getChoice(HttpServletRequest request) {
+    // Get the input from the form.
+    String choiceString = request.getParameter("messageChoice");
+    // Convert the input to an int.
+    int choice;
+    try {
+      choice = Integer.parseInt(choiceString);
+    } catch (NumberFormatException e) {
+      System.err.println("Could not convert to int: " + choiceString);
+      return -1;
```

Comment:
I like the thought put into error handling and bounds checking here. Really awesome. Not trusting user input is one of the top ways to keep a server secure.

Returning values like -1 are called "sentinel values". Values that have special meaning that every time this function is called, they have to be handled. There's been a push lately to try to move away from sentinel values, but you still will see them everywhere. Another way to implement this is to return desired or bounded defaults. If it's an invalid number, you can return a default value like 3. If it's less than 1, you can return 1. Greater than 3, return 3.

A second way to approach this that is beyond what the lessons have shown so far is to set it up to return an error to the user. Something like a sign up form would want to show errors instead of defaults since it's not always possible to know the best default. But in cases like this, falling back to other values makes sense to give a reliable user experience.

Reply:
Looks great!